Firestore/Source/API/FIRDocumentSnapshot.mm

```diff
@@ -187,48 +199,97 @@ - (FieldValueOptions)optionsForServerTimestampBehavior:
   SUPPRESS_END()
 }
 
-- (id)convertedValue:(FSTFieldValue *)value options:(const FieldValueOptions &)options {
-  if (value.type == FieldValue::Type::Object) {
-    return [self convertedObject:(FSTObjectValue *)value options:options];
-  } else if (value.type == FieldValue::Type::Array) {
-    return [self convertedArray:(FSTArrayValue *)value options:options];
-  } else if (value.type == FieldValue::Type::Reference) {
-    const auto &ref = ((FSTDelegateValue *)value).referenceValue;
-    const DatabaseId &refDatabase = ref.database_id();
-    const DatabaseId &database = _snapshot.firestore()->database_id();
-    if (refDatabase != database) {
-      // TODO(b/32073923): Log this as a proper warning.
-      NSLog(@"WARNING: Document %@ contains a document reference within a different database "
-             "(%s/%s) which is not supported. It will be treated as a reference within the "
-             "current database (%s/%s) instead.",
-            self.reference.path, refDatabase.project_id().c_str(),
-            refDatabase.database_id().c_str(), database.project_id().c_str(),
-            database.database_id().c_str());
-    }
-    const DocumentKey &key = ref.key();
-    return [[FIRDocumentReference alloc] initWithKey:key firestore:_snapshot.firestore()];
+- (id)convertedValue:(FieldValue)value options:(const FieldValueOptions &)options {
+  switch (value.type()) {
```

Comment:
This is something I think we should port to the other platforms. Essentially it gets `FieldValue` completely out of the business of needing to implement `value` or `valueWithOptions:`. In the C++ core, this is required because the types of the resulting values depend on the language binding, but I think this ends up just being cleaner.